orttraining/orttraining/python/training/torchdynamo/ort_backend.py

```diff
@@ -0,0 +1,478 @@
+    if first_not_placeholder is None:
+        return
+    for placeholder in placeholders:
+        first_not_placeholder.prepend(placeholder)
+
+
+def _fx_to_torchscript(fx_module: torch.fx.GraphModule) -> torch.jit.ScriptModule:
+    for node in fx_module.graph.nodes:
+        if (
+            node.target == torch.ops.aten._to_copy
+            and len(node.args) == 1
+            and len(node.kwargs) == 1
+            and "dtype" in node.kwargs
+        ):
+            node.target = torch.ops.aten.to
+    for node in fx_module.graph.nodes:
+        new_kwargs = {}
+        for k, v in node.kwargs.items():
+            if isinstance(v, torch.device):
+                v = v.type
+            new_kwargs[k] = v
+        node.kwargs = new_kwargs
+    for node in fx_module.graph.nodes:
+        if isinstance(node.target, torch._ops.OpOverload):
+            node.target = node.target.overloadpacket
+    fx_module.graph.lint()
+    fx_module.recompile()
+    return torch.jit.script(fx_module)
+
+
+def _decorate_script_module(script_module: torch.jit.ScriptModule, expected_inputs, expected_outputs):
+    for i, input_value in enumerate(script_module.graph.inputs()):
+        if input_value.debugName() == "self":
+            script_module.graph.eraseInput(i)
+            break
+    for i, input_value in enumerate(script_module.graph.inputs()):
+        input_value.setType(torch._C.TensorType.create_from_tensor(expected_inputs[i]))
+    for i, output_value in enumerate(script_module.graph.outputs()):
+        output_value.setType(torch._C.TensorType.create_from_tensor(expected_outputs[i]))
+
+
+def _create_onnx_proto(script_module):
+    onnx_proto = _jit_graph_to_onnx_model(script_module.graph, torch.onnx.OperatorExportTypes.ONNX)
+    return onnx_proto
+
+
+def _create_onnx_model(onnx_proto):
+    return onnx.ModelProto.FromString(onnx_proto)
+
+
+def _create_onnx_session(onnx_proto, device):
+    # TODO(wechi): Add more EPs per PyTorch device types.
+    # TODO(wechi): enable external allocators.
+    if device.type == "cuda":
+        return onnxruntime.InferenceSession(onnx_proto, providers=["CUDAExecutionProvider"])
+    return onnxruntime.InferenceSession(onnx_proto, providers=["CPUExecutionProvider"])
+
+
+def _get_onnx_devices(values: Tuple[torch.Tensor]) -> List[ORTC.OrtDevice]:
+    devices: List[ORTC.OrtDevice] = []
+    first_device = values[0].device
+    for value in values:
+        assert value.device == first_device, "All values must be on the same device"
+        device_id = value.device.index or 0
+        devices.append(
+            ORTC.OrtDevice(
+                _get_ort_device_type(value.device.type, device_id), ORTC.OrtDevice.default_memory(), device_id
+            )
+        )
+    return devices
+
+
+def _run_onnx_session_with_ortvaluevector(
+    sess: onnxruntime.InferenceSession,
+    input_names: List[str],
+    inputs: Tuple[torch.Tensor],
+    input_devices: List[ORTC.OrtDevice],
```

Comment:
Humm, I want them to be as specific as possible. Comparing to weakly-typed implementation, I prefer to have strongly-typed implementation. As you may have observed, there is a wave of making Python code strongly-typed, and I'd like to follow. Thanks.

Reply:
I think this more about API design and the contract we make with users (assuming this is public facing api). Making the type more relaxed actually places _more_ constraint on the implementation, because there is less assumption made on the input (e.g. A sequence is not guaranteed to be mutable, like List is). I have learned that it's usually good to be general in input params and be specific in return types. Happy to discuss more.